lib/middlewares/route.js

```diff
@@ -15,18 +15,26 @@ module.exports = function(app) {
     if (method !== 'GET' && method !== 'HEAD') return next();
 
     let url = route.format(decodeURIComponent(req.url));
-    const data = route.get(url);
-    const extname = pathFn.extname(url);
+    let data = route.get(url);
+    let extname = pathFn.extname(url);
 
-    // When the URL is `foo/index.html` but users access `foo`, redirect to `foo/`.
     if (!data) {
-      if (extname) return next();
-
-      url = encodeURI(url);
-      res.statusCode = 302;
-      res.setHeader('Location', `${root + url}/`);
-      res.end('Redirecting');
-      return;
+      if (route.get(url + '/index.html')) {
+        // When the URL is `foo/index.html` but users access `foo`, redirect to `foo/`.
```

Comment:
What if `foo/index.html` & `foo.html` both existed?

Should we prefer `foo.html` before redirect to `foo/index.html`?

Reply:
I assume `foo/index.md` is more common as hexo-server currently doesn't support `foo.md`.

Reply:
I have set up a simple example on Netlify: https://pretty-url.netlify.com/

Here are 4 links:

- https://pretty-url.netlify.com/foo.html
- https://pretty-url.netlify.com/foo
- https://pretty-url.netlify.com/foo/index.html
- https://pretty-url.netlify.com/foo/

Reply:
Here is another example on GitHub Pages: https://github.com/SukkaLab/pretty-url/ https://lab.skk.moe/pretty-url/

- https://lab.skk.moe/pretty-url/foo.html => `/foo.html`
- https://lab.skk.moe/pretty-url/foo => `/foo.html`
- https://lab.skk.moe/pretty-url/foo/index.html => `/foo/index.html`
- https://lab.skk.moe/pretty-url/foo/ => `/foo/index.html`

Reply:
> 
    https://pretty-url.netlify.com/foo.html => /foo.html
    https://pretty-url.netlify.com/foo => /foo.html
    https://pretty-url.netlify.com/foo/index.html => /foo/index.html
    https://pretty-url.netlify.com/foo/ => 301 /foo

`/foo/` redirects to `/foo` which is then proxied to `/foo.html`?

Anyhow, I consider `foo/index.html + foo.html` to be invalid, not an edge case. Besides, `hexo new page` creates `source/foo/index.md` by default. So, if `foo.html` is prioritized, `route.get()` is more likely to be called three times, versus two in `foo/index.html`.

Reply:
@curbengh 

> `/foo/` redirects to `/foo` which is then proxied to `/foo.html`?

So I prefer the GitHub Pages' behavior:

```
/foo/ => /foo/index.html
/foo => /foo.html
```

For hexo-server, when `/foo` is requested, we should find `foo.html` first, then find `/foo/index.html`

Reply:
```js
if (route.get(url + '.html')) {
 // ...
} else if (route.get(url + '/index.html')) {
  // ...

}

Reply:
> For hexo-server, when `/foo` is requested, we should find `foo.html` first, then find `/foo/index.html`

Even if it's the default behavior of gh-pages/apache/nginx, shouldn't we follow hexo's default behavior?

> `hexo new page` creates `source/foo/index.md` by default.

I meant to convey that `source/foo/index.md` should be more common, even though `source/foo.md` is supported as well.

Reply:
@curbengh No matter how common `source/foo/index.md` is, the point is if `source/foo/index.md` and `source/foo.md` both exists, then `source/foo.md` will no longer be accessible through `/foo`, which is not consistent with hexo's current `pretty_url`.

Reply:
I've changed the order:

Location `foo/bar.html` & `foo/bar/index.html`
Request `foo/bar`
Respond `foo/bar.html` via transparent proxy

Location `foo/bar.html` & `foo/bar/index.html`
Request `foo/bar/`
Respond `foo/bar/index.html` via transparent proxy